fedbiomed/researcher/job.py

```diff
@@ -475,22 +497,136 @@ def start_nodes_training_round(self,
+            aux_var=aux_var, nodes=list(self._nodes)
+        )
+        # Upload the shared information that all nodes will download.
+        if aux_shared:
+            path = os.path.join(
+                self._keep_files_dir, f"aux_var_shared_{uuid.uuid4()}.mpk"
+            )
+            Serializer.dump(aux_shared, path)
+            url_shared = self.repo.upload_file(path)["file"]
+        else:
+            url_shared = None
+        # Upload the node-specific information, with node-specific urls.
+        url_bynode = {}  # type: Dict[uuid.UUID, str]
+        for node_id, node_aux in aux_bynode.items():
+            if not node_aux:
+                continue
+            path = os.path.join(
+                self._keep_files_dir,
+                f"aux_var_node_{node_id}_{uuid.uuid4()}.mpk"
+            )
+            Serializer.dump(node_aux, path)
+            url_bynode[node_id] = self.repo.upload_file(path)["file"]
+        # Return the urls of the uploaded files.
+        return url_shared, url_bynode
+
+    @staticmethod
+    def _prepare_agg_optimizer_aux_var(
+        aux_var: Dict[str, Dict[str, Any]],
+        nodes: List[uuid.UUID],
+    ) -> Tuple[
+        Dict[str, Dict[str, Any]],
+        Dict[uuid.UUID, Dict[str, Dict[str, Any]]],
+    ]:
+        """Collect and structure researcher-side Optimizer auxiliary variables.
+
+        Args:
+            aux_var: Auxiliary variables with to structure into multiple dicts,
+                from `{mod_name: (shared_dict | {node_id: node_dict})}` to
+                `{mod_name: shared_dict}` & `{node_id: {mod_name: node_dict}}`.
+            nodes: Ids of the nodes to whom auxiliary variables should be
+                sent. This is used to drop information of non-participating
+                nodes.
+
+        Returns:
+            aux_shared: Dict containing auxiliary variables that are shared
+                across all nodes, with `{mod_name: shared_dict}` format.
+            aux_bynode: Dict containing node-wise dicts of node-specific
+                auxiliary variables, with `{node_id: {mod_name: node_dict}}`
+                format.
+        """
+        aux_shared = {}  # type: Dict[str, Dict[str, Any]]
+        aux_bynode = {}  # type: Dict[uuid.UUID, Dict[str, Dict[str, Any]]]
+        # Iterate over nodes and plug-in-module-wise auxiliary variables.
+        for node_id in nodes:
+            aux_bynode[node_id] = {}
+            for mod_name, mod_info in aux_var.items():
+                # Case of node-specfic information.
+                if node_aux := mod_info.get(str(node_id)):
+                    aux_bynode[node_id][mod_name] = node_aux
+                # Case of global information shared with all nodes.
+                elif mod_name not in aux_shared:
+                    aux_shared[mod_name] = mod_info
+        # Return the restructured auxiliary variables dicts.
+        return aux_shared, aux_bynode
```

Comment:
I was wondering here what would happen if user is using same `OptiModule` twice. Let 's say he is using an optimizer such as `Optimizer(lr=.123, modules=[ScaffoldServer(), ScaffoldServer()])` (which doesnot make sense, but is allowed by declearn I guess). Would the above work ?

Reply:
No, it would not. In general, you are allowed to use multiple instances of the same plug-in, but for auxiliary variables there will be a concurrence, i.e. the last on in the chain will overwrite the variables emitted by others, and be the only one to process received ones.

This may change in the future if we see a use case for it, but at the moment there is none - as you said, implementing two Scaffold controllers at once makes no sense. It is bound to raise an error at some point (that point being the first call to `process_aux_var`), and for us it is really up to the users not to write up things that are mathematically senseless.

Reply:
When we implement a tool to do some early control of configuration correctness, this is typically something that will be caught. In the meanwhile, we may add some check and verbose exception in DecLearn 2.4 - I forgot to do it because in early versions of declearn 2, it was not possible to have multiple instances of the same module.

Reply:
So should we raise an error here ? or should we raise it when creating `Optimizer`? Or should we leave it as it is?

Reply:
I think we should leave it as it is, and tackle it as part of a future task about setting up a controller that would run coherence checks / ad-hoc tests on the Researcher side prior to starting training.

Reply:
OK ! I am going to write a little comment about that, but I am ok to leave it as it is

Reply:
added